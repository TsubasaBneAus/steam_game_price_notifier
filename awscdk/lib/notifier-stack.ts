import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as dotenv from "dotenv";

dotenv.config();

// A stack for the Steam Game Prices Notifier
export class NotifierStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a log group
    const logGroup = new cdk.aws_logs.LogGroup(this, "LogGroup", {
      logGroupName: "steam-game-prices-notifier-log-group",
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      retention: cdk.aws_logs.RetentionDays.ONE_WEEK,
    });

    // Create a Lambda function
    const lambda = new cdk.aws_lambda.Function(this, "Lambda", {
      functionName: "steam-game-prices-notifier-lambda",
      code: cdk.aws_lambda.Code.fromAsset("../function.zip"),
      architecture: cdk.aws_lambda.Architecture.ARM_64,
      runtime: cdk.aws_lambda.Runtime.PROVIDED_AL2023,
      handler: "bootstrap",
      environment: {
        NOTION_API_KEY: process.env.NOTION_API_KEY ?? "",
        NOTION_DATABASE_ID: process.env.NOTION_DATABASE_ID ?? "",
        DISCORD_WEBHOOK_ID: process.env.DISCORD_WEBHOOK_ID ?? "",
        DISCORD_WEBHOOK_TOKEN: process.env.DISCORD_WEBHOOK_TOKEN ?? "",
        STEAM_USER_ID: process.env.STEAM_USER_ID ?? "",
      },
      timeout: cdk.Duration.minutes(2),
      logGroup: logGroup,
      loggingFormat: cdk.aws_lambda.LoggingFormat.JSON,
    });

    // Create a EventBridge rule (UTC)
    const rule = new cdk.aws_events.Rule(this, "Rule", {
      ruleName: "steam-game-prices-notifier-rule",
      schedule: cdk.aws_events.Schedule.cron({
        minute: "0",
        hour: "9",
        day: "*",
        month: "*",
        year: "*",
      }),
    });
    rule.addTarget(new cdk.aws_events_targets.LambdaFunction(lambda));

    // Create a OIDC provider
    const provider = new cdk.aws_iam.OpenIdConnectProvider(
      this,
      "OIDCProvider",
      {
        url: "https://token.actions.githubusercontent.com",
        clientIds: ["sts.amazonaws.com"],
      }
    );

    // Create a role for the OIDC provider
    const role = new cdk.aws_iam.Role(this, "Role", {
      roleName: "steam-game-prices-notifier-github-actions-role",
      assumedBy: new cdk.aws_iam.WebIdentityPrincipal(
        provider.openIdConnectProviderArn,
        {
          StringEquals: {
            "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
          },
          StringLike: {
            "token.actions.githubusercontent.com:sub":
              "repo:TsubasaBneAus/steam_game_prices_notifier:*",
          },
        }
      ),
    });
    role.addToPolicy(
      new cdk.aws_iam.PolicyStatement({
        actions: ["lambda:UpdateFunctionCode"],
        resources: [lambda.functionArn],
      })
    );
  }
}
