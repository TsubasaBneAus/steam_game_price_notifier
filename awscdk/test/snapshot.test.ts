import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { NotifierStack } from "../lib/notifier-stack";

test("Snapshot test", () => {
  const app = new cdk.App();
  const stack = new NotifierStack(app, "NotifierStack");
  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
