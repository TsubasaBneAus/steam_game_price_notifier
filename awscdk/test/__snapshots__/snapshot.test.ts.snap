// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": {
      "DependsOn": [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2926656cdb97b26d98f8b957b0d3f7326b458679745f2817e50333a21767350c.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaD247545B": {
      "DependsOn": [
        "LambdaServiceRoleA8ED4D3B",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "5c730d23bf6eb83d48b3ada65078c477a9dbdf7af4233de830d9e4874c0e10ad.zip",
        },
        "Environment": {
          "Variables": {
            "DISCORD_WEBHOOK_ID": "1310250977953185792",
            "DISCORD_WEBHOOK_TOKEN": "6-TH62G_tTLSMm4YlE6scLzyXmhLqZVRelDrV8MGvBnIhEPXINidSzu2sYslvdJfzAOm",
            "NOTION_API_KEY": "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",
            "NOTION_DATABASE_ID": "1464ff0d00e9802d9275eced7bcb8c30",
            "STEAM_USER_ID": "76561199013163152",
          },
        },
        "FunctionName": "steam-game-prices-notifier-lambda",
        "Handler": "bootstrap",
        "LoggingConfig": {
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "LogGroupF5B46931",
          },
        },
        "Role": {
          "Fn::GetAtt": [
            "LambdaServiceRoleA8ED4D3B",
            "Arn",
          ],
        },
        "Runtime": "provided.al2023",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaServiceRoleA8ED4D3B": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogGroupF5B46931": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": "steam-game-prices-notifier-log-group",
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "OIDCProviderA3376E13": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ClientIDList": [
          "sts.amazonaws.com",
        ],
        "CodeHash": "2926656cdb97b26d98f8b957b0d3f7326b458679745f2817e50333a21767350c",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "Role1ABCC5F0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "StringEquals": {
                  "token.actions.githubusercontent.com:aud": "sts.amazonaws.com",
                },
                "StringLike": {
                  "token.actions.githubusercontent.com:sub": "repo:TsubasaBneAus/steam_game_prices_notifier:*",
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": {
                  "Ref": "OIDCProviderA3376E13",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": "steam-game-prices-notifier-github-actions-role",
      },
      "Type": "AWS::IAM::Role",
    },
    "RoleDefaultPolicy5FFB7DAB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:UpdateFunctionCode",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "LambdaD247545B",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "RoleDefaultPolicy5FFB7DAB",
        "Roles": [
          {
            "Ref": "Role1ABCC5F0",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Rule4C995B7F": {
      "Properties": {
        "Name": "steam-game-prices-notifier-rule",
        "ScheduleExpression": "cron(0 9 * * ? *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::GetAtt": [
                "LambdaD247545B",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "RuleAllowEventRuleNotifierStackLambdaBD511012C866015B": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "LambdaD247545B",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": {
          "Fn::GetAtt": [
            "Rule4C995B7F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
